cmd/gitsign-attest/internal/attest/attest.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2022 The Sigstore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attest
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/filemode"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage"
+	"github.com/sigstore/cosign/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/cmd/cosign/cli/sign"
+	"github.com/sigstore/cosign/pkg/cosign"
+	"github.com/sigstore/cosign/pkg/cosign/attestation"
+	"github.com/sigstore/cosign/pkg/types"
+	rekorclient "github.com/sigstore/rekor/pkg/generated/client"
+	dssesig "github.com/sigstore/sigstore/pkg/signature/dsse"
+	signatureoptions "github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+const (
+	CommitRef = "refs/attestations/commits"
+	TreeRef   = "refs/attestations/trees"
+)
+
+// WriteFile writes the given file + a DSSE signed attestation to the corresponding attestation ref.
+// The SHA of the created commit is returned.
+func WriteFile(ctx context.Context, repo *git.Repository, refName string, sha plumbing.Hash, path, attType string) (plumbing.Hash, error) {
+	b, err := os.ReadFile(path)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Write the blob we received verbatim.
+	// TODO: is this necessary? should we just extract this data from DSSE?
+	blobHash, err := writeBlob(repo.Storer, b)
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	// Step 1: Write the files
+
+	// Create the DSSE, sign it, store it.
+	sig, err := signPayload(ctx, sha, b, attType)
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+	sigHash, err := writeBlob(repo.Storer, sig)
+	if err != nil {
+		return plumbing.ZeroHash, err
+	}
+
+	// Create 2 files: 1 mirroring the original file basename,
+	// another using <basename>.sig for the DSSE.
+	// TODO: prevent accidental file overwrites.
```

Comment:
How should we prevent them, just detecting if the file already exists and fail? Is that something we can squeeze into this PR?